src/test/java/org/opentripplanner/middleware/controllers/api/TrackedTripControllerTest.java

```diff
@@ -504,6 +509,46 @@ void canNotUpdateCompletedJourney() throws Exception {
         assertEquals(HttpStatus.BAD_REQUEST_400, response.status);
     }
 
+    @Test
+    void canRerouteTrip() throws Exception {
+        assumeTrue(IS_END_TO_END);
+
+        StartTrackingPayload startTrackingPayload = createStartTrackingPayload(multipleItinMonitoredTrip.id);
+
+        // Start tracking.
+        var response = makeRequest(
+            START_TRACKING_TRIP_PATH,
+            JsonUtils.toJson(startTrackingPayload),
+            headers,
+            HttpMethod.POST
+        );
+
+        var startTrackingResponse = JsonUtils.getPOJOFromJSON(response.responseBody, TrackingResponse.class);
+        trackedJourney = Persistence.trackedJourneys.getById(startTrackingResponse.journeyId);
+        assertEquals(HttpStatus.OK_200, response.status);
+
+        // Make call directly to reroute trip, bypassing call to OTP instance.
+        assertTrue(ManageTripTracking.rerouteTrip(
+            new TripTrackingData(multipleItinMonitoredTrip, trackedJourney, trackedJourney.locations), true)
+        );
+        TrackedJourney updated = Persistence.trackedJourneys.getById(trackedJourney.id);
+        assertEquals(1, updated.reroutings.size());
+        assertEquals(-1, updated.longestConsecutiveDeviatedPoints);
+
+        MonitoredTrip trip = Persistence.monitoredTrips.getById(multipleItinMonitoredTrip.id);
+        assertEquals(trip.itinerary.duration, trip.journeyState.matchingItinerary.duration);
+        assertEquals(trip.itinerary.legs.size(), trip.journeyState.matchingItinerary.legs.size());
+
+        // Reroute again from a different location.
+        trackedJourney.locations.clear();
+        trackedJourney.locations.add(new TrackingLocation(180, 28.45119,-81.36818, 3, getDateAndConvertToSeconds()));
+        assertTrue(ManageTripTracking.rerouteTrip(
+            new TripTrackingData(multipleItinMonitoredTrip, trackedJourney, trackedJourney.locations), true)
+        );
+        updated = Persistence.trackedJourneys.getById(trackedJourney.id);
+        assertEquals(2, updated.reroutings.size());
+        assertEquals(-1, updated.longestConsecutiveDeviatedPoints);
```

Comment:
There should be an assertion that the "from" of the rerouted itinerary should be the same as the last location in `trackedJourney`.